## Daniel Pastrana Vargas

### 733982
#### Ingeniería Financiera

# Tarea - Programación lineal

## Problema 1

### Diseño de la Dieta Óptima

Se quiere producir comida para gatos de la manera más barata, no obstante se debe también asegurar que se cumplan los datos requeridos de analisis nutricional. Por lo que se quiere variar la cantidad de cada ingrediente para cumplir con los estandares nutricionales. Los requisitos que se tienen es que en $100$ gramos, se deben tener **por lo menos** $8$ gramos de proteína y $6$ gramos de grasa. Así mismo, no se debe tener más de $2$ gramos de fibra y $0.4$ gramos de sal.  

Los datos nutricionales se pueden obtener de la siguiente tabla:

Ingrediente|Proteína|Grasa|Fibra|Sal
:----|----|----|----|----
Pollo|  10.0%|08.0%|00.1%|00.2%
Carne|  20.0%|10.0%|00.5%|00.5%
Cordero|15.0%|11.0%|00.5%|00.7%
Arroz|  00.0%|01.0%|10.0%|00.2%
Trigo|  04.0%|01.0%|15.0%|00.8%
Gel|    00.0%|00.0%|00.0%|00.0%

Los costos de cada producto son:

Ingrediente|Costo por gramo
:----|----
Pollo|$\$$0.013
Carne|$\$$0.008
Cordero|$\$$0.010
Arroz|$\$$0.002
Trigo|$\$$0.005
Gel|$\$$0.001    

Lo que se busca optimizar en este caso es la cantidad de productos que se debe utilizar en la comida de gato, minimizando el costo total. Para simplificar la notación use las siguientes variables: 

+ $x_1:$ Gramos de pollo  
+ $x_2:$ Gramos de carne  
+ $x_3:$ Gramos de cordero  
+ $x_4:$ Gramos de arroz  
+ $x_5:$ Gramos de trigo  
+ $x_6:$ Gramos de gel

La tarea consiste en plantear el problemade programación lineal que permita satisfacer las necesidades alimenticias del gato al tiempo que minimice el costo total y resolverlo con `linprog`.

### Paso 1. Criterio de optimización:
Hay que minimizar el costo de la comida del gato.

### Paso 2. Variables que afectan mi criterio de optimización y tengo control:
Cantidad de producto en gramos que compro $(x_1, x_2, x_3, x_4, x_5, x_6)$.

### Paso 3. Función de costo:
$$ f(x_1...x_6) = 0.013x_1 + 0.008x_2 + 0.010x_3 + 0.002x_4 + 0.005x_5 + 0.001x_6 $$

### Paso 4. Restricciones:
$$ 0.100x_1 + 0.200x_2 + 0.150x_3 + 0.040x_5 \geq 8 $$
$$ 0.080x_1 + 0.100x_2 + 0.110x_3 + 0.010x_4 + 0.010x_5 \geq 6 $$
$$ 0.001x_1 + 0.005x_2 + 0.005x_3 + 0.100x_4 + 0.150x_5 \leq 2 $$
$$ 0.002x_1 + 0.005x_2 + 0.007x_3 + 0.002x_4 + 0.008x_5 \leq 0.4 $$

In [2]:
import scipy.optimize as opt
import numpy as np

c = np.array([-0.013,
              -0.008,
              -0.010,
              -0.002, 
              -0.005,
             ])

Aub = np.array([
    [-0.100, -0.200, -0.150, 0.000, -0.040],
    [-0.080, -0.100, -0.110, -0.010, -0.010],
    [0.001, 0.005, 0.005, 0.100, 0.150],
    [0.002, 0.005, 0.007, 0.002, 0.008]
               ])

bub = np.array([-8.0,
                -6.0,
                2.0,
                0.4
               ])

opt.linprog (c, Aub, bub)

           con: array([], dtype=float64)
 crossover_nit: 0
         eqlin:  marginals: array([], dtype=float64)
  residual: array([], dtype=float64)
           fun: -2.6000000000000005
       ineqlin:  marginals: array([-0. , -0. , -0. , -6.5])
  residual: array([12. , 10. ,  1.8,  0. ])
         lower:  marginals: array([0.    , 0.0245, 0.0355, 0.011 , 0.047 ])
  residual: array([200.,   0.,   0.,   0.,   0.])
       message: 'Optimization terminated successfully. (HiGHS Status 7: Optimal)'
           nit: 4
         slack: array([12. , 10. ,  1.8,  0. ])
        status: 0
       success: True
         upper:  marginals: array([0., 0., 0., 0., 0.])
  residual: array([inf, inf, inf, inf, inf])
             x: array([200.,   0.,   0.,   0.,   0.])

La forma más barata de hacer la comida de gato, es usando 200 gramos de pollo, y nada más. El costo es de $2.6

## Problema 2

### Problema de distribución de energía eléctrica

La Comisión Federal de Electricidad **(CFE)** dispone de cuatro plantas de generación para satisfacer la demanda diaria eléctrica en cuatro ciudades, Guadalajara, León y Morelia. Las plantas $1$, $2$ y $3$ pueden satisfacer $80$, $40$ y $60$ millones de KW al día respectivamente. Las necesidades de las ciudades de Guadalajara, León y Morelia son de $70$, $40$ y $70$ millones de Kw al día respectivamente. 


Los costos asociados al envío de suministro energético por cada millón de Kw entre cada planta y cada ciudad son los registrados en la siguiente tabla. 

-|Guadalajara|León|Morelia
:----|----|----|----
Planta 1|5|2|7
Planta 2|3|6|6
Planta 3|6|1|2

Y por último, las restricciones del problema, van a estar dadas por las capacidades de oferta y demanda de cada planta (en millones de KW) y cada ciudad.

Para simplificar la notación use las siguientes variables:

+ $x_1$: Kw (en millones) distribuidos de la Planta 1 a Guadalajara
+ $x_2$: Kw (en millones) distribuidos de la Planta 1 a León
+ $x_3$: Kw (en millones) distribuidos de la Planta 1 a Morelia
+ $x_4$: Kw (en millones) distribuidos de la Planta 2 a Guadalajara
+ $x_5$: Kw (en millones) distribuidos de la Planta 2 a León
+ $x_6$: Kw (en millones) distribuidos de la Planta 2 a Morelia
+ $x_7$: Kw (en millones) distribuidos de la Planta 3 a Guadalajara
+ $x_8$: Kw (en millones) distribuidos de la Planta 3 a León
+ $x_9$: Kw (en millones) distribuidos de la Planta 3 a Morelia

La tarea consiste en plantear el problema de programación lineal que permita satisfacer las necesidades de todas las ciudades al tiempo que minimice los costos asociados a la distribución y resolverlo con `linprog`.

### Paso 1. Criterio de optimización:
Hay que minimizar los costos de distribución.

### Paso 2. Variables que afectan mi criterio de optimización y tengo control:
Cantidad de producto en gramos que compro $(x_{AD}, x_{AE}, x_{AF}, x_{BD}, x_{BE}, x_{BF}, x_{CD}, x_{CE}, x_{CF})$.

### Paso 3. Función de costo:
$$ f(x_{AD}...x_{CF}) = 5x_{AD} + 2x_{AE} + 7x_{AF} + 3x_{BD} + 6x_{BE} + 6x_{BF} + 6x_{CD} + x_{CE} + 2x_{CF} $$

### Paso 4. Restricciones:
Restricciones de la producción de kw:
$$x_{AD} + x_{AE} + x_{AF} \leq 80$$
$$x_{BD} + x_{BE} + x_{BF} \leq 40$$
$$x_{CD} + x_{CE} + x_{CF} \leq 60$$

Restricciones de la necesidad de kw:
$$x_{AD} + x_{BD} + x_{CD} \geq 70$$
$$x_{AE} + x_{BE} + x_{CE} \geq 40$$
$$x_{AF} + x_{BF} + x_{CF} \geq 70$$

In [3]:
import scipy.optimize as opt
import numpy as np

c = np.array([5,
              2,
              7,
              3, 
              6,
              6,
              6,
              1,
              2
             ])

Aub = np.array([
    [1, 1, 1, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 1, 1, 1, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 1, 1, 1]
               ])

bub = np.array([80,
                40,
                60
               ])

Aeq = np.array([
    [1, 0, 0, 1, 0, 0, 1, 0, 0],
    [0, 1, 0, 0, 1, 0, 0, 1, 0],
    [0, 0, 1, 0, 0, 1, 0, 0, 1]
])

beq = np.array([70,
                40,
                70
               ])

opt.linprog (c, Aub, bub, Aeq, beq)

           con: array([0., 0., 0.])
 crossover_nit: 0
         eqlin:  marginals: array([5., 2., 7.])
  residual: array([0., 0., 0.])
           fun: 540.0
       ineqlin:  marginals: array([-0., -2., -5.])
  residual: array([0., 0., 0.])
         lower:  marginals: array([0., 0., 0., 0., 6., 1., 6., 4., 0.])
  residual: array([30., 40., 10., 40.,  0.,  0.,  0.,  0., 60.])
       message: 'Optimization terminated successfully. (HiGHS Status 7: Optimal)'
           nit: 6
         slack: array([0., 0., 0.])
        status: 0
       success: True
         upper:  marginals: array([0., 0., 0., 0., 0., 0., 0., 0., 0.])
  residual: array([inf, inf, inf, inf, inf, inf, inf, inf, inf])
             x: array([30., 40., 10., 40.,  0.,  0.,  0.,  0., 60.])

El costo mínimo es de 540.
- La planta 1 debe de mandar 30 millones de kw a Guadalajara, 40 millones de kw a León y 10 millones de kw a Morelia.
- La planta 2 debe de mandar 40 millones de kw a Guadalajara, 0 millones de kw a León y 0 millones de kw a Morelia.
- La planta 3 debe de mandar 0 millones de kw a Guadalajara, 0 millones de kw a León y 60 millones de kw a Morelia.